kvstore: Add transaction support to KV storage

chaudyg opened the pull request:
Introduced transaction operations with compare-and-swap semantics.

This update enhances the KV interface, allowing for more complex data manipulation while ensuring data integrity through transactional guarantees.

chaudyg committed:
Add transaction support to KV storage

- Introduced transaction operations with compare-and-swap semantics.
- Added types and functions for comparisons (e.g., CompareExists, CompareValue) and transaction operations (e.g., TxnOpPut, TxnOpDelete).
- Implemented Txn method to execute transactions based on comparisons and success/failure operations.
- Added validation for transaction requests to enforce limits on comparisons and operations.
- Enhanced testing suite to cover various transaction scenarios, including success and failure cases, and edge conditions.

This update enhances the KV interface, allowing for more complex data manipulation while ensuring data integrity through transactional guarantees.

gassiss commented on pkg/storage/unified/resource/kv.go:
```diff
@@ -35,6 +35,106 @@ type ListOptions struct {
+// CompareKeyExists creates a comparison that checks if a key exists or not.
+// If exists is true, the comparison succeeds if the key exists.
+// If exists is false, the comparison succeeds if the key does not exist.
+func CompareKeyExists(key string, exists bool) Compare {
```
having a separate one `CompareKeyNotExists` and hard-coding the `exists` param for each function would be easier on the eyes IMO

gassiss commented on pkg/storage/unified/resource/kv.go:
```diff
@@ -360,3 +465,106 @@ func (k *badgerKV) BatchDelete(ctx context.Context, section string, keys []strin
+					succeeded = false
+				}
+			} else {
+				item, err := txn.Get([]byte(keyWithSection))
```
why not save the item in line 485?

gassiss commented on pkg/storage/unified/testing/kv.go:
```diff
@@ -801,3 +803,357 @@ func saveKVHelper(t *testing.T, kv resource.KV, ctx context.Context, section, ke
+			{Type: resource.TxnOpPut, Key: "created-key", Value: []byte("created-value")},
+		}
+
+		resp, err := kv.Txn(ctx, section, cmps, successOps, nil)
```
I'd put some dummy operations instead of `nil` and check that it wasn't executed like you did in some other tests

chaudyg committed:
Enhance comparison functions in KV storage

- Updated the CompareKeyExists function to simplify its usage by removing the 'exists' parameter, making it always succeed if the key exists.
- Introduced a new CompareKeyNotExists function to check for non-existent keys.
- Modified transaction tests to utilize the new comparison functions, ensuring accurate behavior for both existing and non-existing keys.

chaudyg commented:
Closing in favour of https://github.com/grafana/grafana/pull/116073

chaudyg commented:
Won't do
